### ISSUE TITLE
Limiting number of cpus used?

### ISSUE THREAD

**Issue by ivirshup** [lmcinnes/umap#208]
I've been noticing that when I run UMAP part of the computation runs in parallel. I'd like to be able to limit the number of processes that get used per transformation. This would be useful for keeping a decent cpu utilization rate without risking getting my jobs killed.

I've tried limiting the number of threads numba uses (`numba.config.NUMBA_NUM_THREADS`), but that hasn't seemed to work.

Is this something which is currently configurable, or a feature that could be added?

Thanks!

**Comment by lmcinnes** [comment]
In practice I believe almost all the actual parallel work is actually done with in ARPACK in a call to ``eigsh``, and unfortunately I am not sure how to enforce that to use fewer cores easily. The most trivial approach would be to avoid using that section of code by setting ``init='random'``. That will lose some stability in the embedding, but should avoid the many core aspect.

**Comment by ivirshup** [comment]
I just gave that a shot, but still saw a bunch of new processes starting up. I've even tried just disabling numba but still get the same behavior. For reproducibilities sake, here's the code I've been testing with and the environment variables I've tried.

```python
# In bash, previously
# export NUMBA_NUM_THREADS=1
# export NUMBA_DISABLE_JIT=1
from sklearn.datasets import load_iris, load_digits
import umap

digits = load_digits()
reducer = umap.UMAP(init="random")
embedding = reducer.fit_transform(digits.data)
```

**Comment by lmcinnes** [comment]
With no numba, and no spectral initialisation you've eliminated most of the options for threading. In general threading is only going to happen in the case of library calls -- everything in umap itself relies on numba for any threading. The most likely candidate for what is going on here is that because you have very little data umap is using ``pairwise_distances`` from sklearn, which I believe can and potentially does parallelise the work. I don't have any explicit options set, and I can't say I know how to turn that off. With larger datasets (more than 4096 samples) this will not be the case.

**Comment by ivirshup** [comment]
Thanks for the help! Running with a larger dataset, the amount of time with multiple cpus has gone down. It looks like there's something else spawning a bunch of python processes for a short period of work. I haven't quite nailed down what that is – but I suspect scikit-learn. I'm going to investigate that a bit more and will hopefully come back with a configuration for limiting the cpus used.

**Comment by sleighsoft** [comment]
Can this be closed?

**Comment by sleighsoft** [comment]
@ivirshup Any updates?

**Comment by wangvg** [comment]
@ivirshup seems to be correct. scikit-learn relies on numpy/scipy routines which are creating multiple processes. What appears to work is setting an environment variable to force the underlying parallelism libraries to use a specific number of processors (per the source below). If scikit-learn was retrieved from pypi (or conda-forge), you need to export the environment variable OPENBLAS_NUM_THREADS, but if it was retrieved from base conda, the environment variable should be MKL_NUM_THREADS (I have not tested this one though). Hopefully this helps anybody else that ran into this issue as well.

[Source (scikit-learn)](https://scikit-learn.org/stable/modules/computing.html#parallelism)

**Comment by lmcinnes** [comment]
In newer versions of umap you may also want to add ``NUMBA_NUM_THREADS`` to the list. There is a pull request underway that would allow user set at runtime thread limits for numba (numba/numba#4615), but it is not yet complete. When that gets added to numba I'll try to update umap to take advantage of that and allow easier user specified cpu usage.

**Comment by yusuftalhatamer** [comment]
@lmcinnes any updates on cpu core limiting?

**Comment by lmcinnes** [comment]
Yes, that pull request got merged and the ``n_jobs`` parameter now makes use of it, so it *should* work if you set it.
